lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+
+        if (! $width && ! $height) {
+            if ($layout === Layout::FILL) {
+                list($width, $height) = $this->getDimensionsFromParent($element);
+            } else {
+                return null;
+            }
+        }
+
+        if ($this->isTinyElement($layout, $width, $height)) {
+            return null;
+        }
+
+        $srcset = $element->getAttribute(Attribute::SRCSET);
+        $media  = $element->getAttribute(Attribute::MEDIA);
+
+        return new HeroImage($src, $media, $srcset, $element);
+    }
+
+    /**
+     * Get the dimensions to use from an element's parent(s).
+     *
+     * @param DOMElement $element Element to check the parents of.
+     * @return array Array containing the width and the height.
+     */
+    private function getDimensionsFromParent(DOMElement $element)
+    {
+        while ($element->parentNode) {
+            $element = $element->parentNode;
+
+            if (! $element instanceof DOMElement) {
+                continue;
+            }
+
+            $width  = $element->getAttribute(Attribute::WIDTH);
+            $height = $element->getAttribute(Attribute::HEIGHT);
+
+            if (! $width && ! $height) {
+                continue;
+            }
+
+            return [$width, $height];
+        }
+
+        return [0, 0];
+    }
+
+    /**
+     * Check whether an element is to be considered tiny and should be ignored.
+     *
+     * A tiny element is any element with width or height less than 150 pixels and a non-responsive layout.
+     *
+     * @param string $layout Layout of the element.
+     * @param int    $width  Width of the element.
+     * @param int    $height Height of the element.
+     * @return bool Whether the element is tiny.
+     */
+    private function isTinyElement($layout, $width, $height)
+    {
+        if ($width <= 0 || $height <= 0) {
+            return true;
+        }
+
+        if ($layout === Layout::INTRINSIC || $layout === Layout::RESPONSIVE) {
+            return false;
+        }
+
+        return $width < self::TINY_IMG_THRESHOLD || $height < self::TINY_IMG_THRESHOLD;
+    }
+
+    /**
+     * Detect a hero image candidate from a video's poster (= placeholder) image.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForPosterImage(DOMElement $element)
+    {
+        $poster = $element->getAttribute(Attribute::POSTER);
+
+        if (! $poster) {
+            return null;
+        }
+
+        if (! Url::isValidImageSrc($poster)) {
+            return null;
+        }
+
+        $width  = $element->getAttribute(Attribute::WIDTH);
+        $height = $element->getAttribute(Attribute::HEIGHT);
+        $layout = $element->getAttribute(Attribute::LAYOUT);
+
+        if ($this->isTinyElement($layout, $width, $height)) {
```

Comment:
What if `layout` is `fixed-height`?

Reply:
I opted to make it only check the height but not the width in that case.